Infra/self hosted runners

DaMandal0rian opened the pull request:
This PR is for the deployment of self-hosted runners with AWS EC2 instances and EKS kubernetes.

- Support for deployments on single stack instances using terraform and AWS EC2
- Currently support for Linux[x86_64, ARM64] , MacOS[x86_64, ARM64] and Windows x86_64
- Linking issue https://github.com/subspace/infra/issues/114

DaMandal0rian committed:
github linux runners helm/kubernetes and aws terraform

DaMandal0rian committed:
linting

DaMandal0rian committed:
add support for mac and windows servers, improve on linux servers declarations

DaMandal0rian committed:
mask tokens

DaMandal0rian committed:
remove certificate file

vedhavyas commented on github-runners/k8s/dockerfile:
```diff
@@ -0,0 +1,34 @@
+COPY --chown=github:github entrypoint.sh runsvc.sh ./
+RUN sudo chmod u+x ./entrypoint.sh ./runsvc.sh
+
+ENTRYPOINT ["/home/github/entrypoint.sh"]
```
Configure your IDE to put a new line in the end pls

nazar-pc replied:
Files should have new line at the end, please configure IDE to insert them automatically

vedhavyas commented on github-runners/k8s/helm/NOTES.txt:
```diff
@@ -0,0 +1,16 @@
+Release "actions-runner-controller" does not exist. Installing it now.
+NAME: actions-runner-controller
+LAST DEPLOYED: Fri Apr 14 00:30:53 2023
+NAMESPACE: actions-runner-system
```
I'm not understanding much from the Notes especially helm stuff. Do they need to be here or Is that a response from the helm upgrade command ?

vedhavyas commented on github-runners/k8s/templates/base/runner.yaml:
```diff
@@ -0,0 +1,34 @@
+      labels:
+        app: k8s-github-runner
+    spec:
+      repository: https://github.com/subspace/subspace-cli
```
```suggestion
      repository: https://github.com/subspace/<REPOSITORY>
```

Prefer  not to default in template as its easier to assume cli is a requirement here when I looked at it initially

nazar-pc commented on github-runners/k8s/dockerfile:
```diff
@@ -0,0 +1,34 @@
+FROM debian:buster-slim
```
Why not Ubuntu though?

nazar-pc commented on github-runners/k8s/dockerfile:
```diff
@@ -0,0 +1,34 @@
+    && rm -rf /var/lib/apt/lists/* \
+    && useradd -m github \
+    && usermod -aG sudo github \
+    && echo "%sudo ALL=(ALL) NOPASSWD:ALL" >> /etc/sudoers
```
If you want me to review this you'd have to explain why we want sudo in the container at all, makes very little sense to me.

DaMandal0rian replied:
Thats the official github implementation, which requires running as sudo user. But I have removed the k8s support so don't need this anymore.

nazar-pc commented on github-runners/k8s/dockerfile:
```diff
@@ -0,0 +1,34 @@
+    && sudo ./bin/installdependencies.sh
+
+COPY --chown=github:github entrypoint.sh runsvc.sh ./
+RUN sudo chmod u+x ./entrypoint.sh ./runsvc.sh
```
Is this the only reason we add sudo? If so this step should happen before we change user. Also I'm not quite sure while file needs to be owned by `github` user, is that user really supposed to modify the file it is running with? Otherwise default permissions should still allow to run it (if you commit `entrypoint.sh` into the repo with `+x` permissions). This looks really unnecessary to me.

DaMandal0rian replied:
Thats the official github implementation, which requires running as sudo user. But I have removed the k8s support so don't need this anymore.

nazar-pc commented on github-runners/k8s/entrypoint.sh:
```diff
@@ -0,0 +1,65 @@
+    TO_BE_INSTALLED=$(echo ${ADDITIONAL_PACKAGES} | tr "," " " )
+    echo "Installing additional packages: ${TO_BE_INSTALLED}"
+    sudo apt-get update && sudo apt-get install -y ${TO_BE_INSTALLED} && sudo apt-get clean
+fi
```
Why? Shouldn't this happen during container building rather than installing?

nazar-pc commented on github-runners/k8s/entrypoint.sh:
```diff
@@ -0,0 +1,65 @@
+fi
+
+registration_url="https://github.com/${GITHUB_OWNER}"
+token_url="https://api.github.com/orgs/${GITHUB_OWNER}/actions/runners/registration-token"
```
Should owner really be compiled into the image? It would make more sense to have it passed as an argument

nazar-pc commented on github-runners/k8s/dockerfile:
```diff
@@ -0,0 +1,34 @@
+FROM debian:buster-slim
+
+ENV GITHUB_PAT ""
+ENV GITHUB_TOKEN ""
```
These look like a very bad idea. Please don't compile tokens into container images and available to everything running it it, they should be passed in as arguments.

nazar-pc commented on github-runners/k8s/helm/helm.sh:
```diff
@@ -0,0 +1,6 @@
+#!/bin/sh
+
+#export GITHUB_TOKEN=$1
+echo $GITHUB_TOKEN
```
Please don't do that

nazar-pc commented on github-runners/k8s/helm/helm.sh:
```diff
@@ -0,0 +1,6 @@
+#export GITHUB_TOKEN=$1
+echo $GITHUB_TOKEN
+kubectl create secret generic controller-manager  -n actions-runner-system --from-literal=github_token=${GITHUB_TOKEN}
+helm upgrade --install --namespace actions-runner-system  --create-namespace --wait actions-runner-controller actions-runner-controller/actions-runner-controller --set syncPeriod=1m
```
Not sure what this is. Why would we create secret manually for instance?

nazar-pc commented on github-runners/k8s/templates/base/acme-certificate.yaml:
```diff
@@ -0,0 +1,67 @@
+kind: ClusterIssuer
+metadata:
+  name: letsencrypt-production
+  namespace: explorer-network
```
Strange namespace `explorer-network`. Also do we need most of this commented-out file?

nazar-pc commented on github-runners/k8s/templates/base/runner.yaml:
```diff
@@ -0,0 +1,34 @@
+          valueFrom:
+            secretKeyRef:
+              name: my-secret
+              key: pat
```
Why do we need both tokens? I don't think PAT should be used for org at all.

nazar-pc commented on github-runners/terraform/linux/main.tf:
```diff
@@ -0,0 +1,83 @@
+    values = ["hvm"]
+  }
+
+  owners = ["099720109477"] # Canonical
```
If variables are possible here, it would be better than a comment :slightly_smiling_face:

DaMandal0rian replied:
The owner is constant, that is the canonical ID. Not sure why a variable is needed here.

nazar-pc commented on github-runners/terraform/linux/main.tf:
```diff
@@ -0,0 +1,83 @@
+}
+
+
+# resource "aws_instance" "web-server" {
```
Is this and other commented-out code really necessary here?

nazar-pc commented on github-runners/terraform/linux/variables.tf:
```diff
@@ -0,0 +1,80 @@
+variable "instance_type" {
+  default = "t2.micro"
+  type    = string
+}
```
I don't think we'll be able to build anything on this, or is it just for testing purposes?

DaMandal0rian replied:
This was for testing, and it's just set in default. We use terraform.tfvars file to set the instance.

nazar-pc commented on github-runners/terraform/base/ami.tf:
```diff
@@ -0,0 +1,102 @@
+data "aws_ami" "ubuntu_x86" {
```
Judging by how images are named, this would be more appropriate:
```suggestion
data "aws_ami" "ubuntu_amd64" {
```
Or at least this:
```suggestion
data "aws_ami" "ubuntu_x86_64" {
```

nazar-pc commented on github-runners/terraform/base/ami.tf:
```diff
@@ -0,0 +1,102 @@
+    values = ["hvm"]
+  }
+
+  owners = ["801119661308"] # Canonical
```
Canonical images for macOS and Windows, really?

DaMandal0rian replied:
this was an old draft already fixed

nazar-pc commented on github-runners/terraform/base/main.tf:
```diff
@@ -0,0 +1,440 @@
+  sudo apt update -y
+  sudo apt upgrade -y
+  sudo apt install git curl jq wget gnupg openssl make build-essential jq net-tools -y
+  sudo curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- --default-toolchain nightly -y
```
Why do we install `jq`, `rustup` and a bunch of other tool here> Wouldn't setup below cover everything runner itself needs?

DaMandal0rian replied:
I've removed jq and rust will be installed via the pipeline during the job run, so removed that. I removed make and build-essentials.

nazar-pc commented on github-runners/terraform/base/main.tf:
```diff
@@ -0,0 +1,440 @@
+  sudo ./svc.sh install
+  sudo ./svc.sh start
+  sudo ./svc.sh status
+  sudo wget -O /tmp/netdata-kickstart.sh https://my-netdata.io/kickstart.sh && sh /tmp/netdata-kickstart.sh --non-interactive --nightly-channel --claim-token ${var.netdata_token} --claim-url https://app.netdata.cloud
```
Why nightly here?

DaMandal0rian replied:
this is to get the latest netdata bug and security fixes.

nazar-pc commented on github-runners/terraform/base/main.tf:
```diff
@@ -0,0 +1,440 @@
+  sudo curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- --default-toolchain nightly -y
+  mkdir actions-runner && cd actions-runner
+  curl -o actions-runner-linux-x64-2.303.0.tar.gz -L https://github.com/actions/runner/releases/download/v2.303.0/actions-runner-linux-x64-2.303.0.tar.gz
+  echo 'e4a9fb7269c1a156eb5d5369232d0cd62e06bec2fd2b321600e85ac914a9cc73  actions-runner-linux-x64-2.303.0.tar.gz' | shasum -a 256 -c
```
I'm wondering if these should be variables too, looks like we'll want to update them in the future.

DaMandal0rian replied:
these have moved to variables

nazar-pc commented on github-runners/terraform/base/main.tf:
```diff
@@ -0,0 +1,440 @@
+  curl -o actions-runner-osx-x64-2.303.0.tar.gz -L https://github.com/actions/runner/releases/download/v2.303.0/actions-runner-osx-x64-2.303.0.tar.gz
+  echo '8bd595568ceee5eb25576972bc8075b47c149b3fac7eb7873deed67944b45739  actions-runner-osx-x64-2.303.0.tar.gz' | shasum -a 256 -c
+  tar xzf ./actions-runner-osx-x64-2.303.0.tar.gz
+  ./config.sh --url https://github.com/subspace --token ${var.GH_TOKEN} --unattended --name mac_x86_64 --labels 'self-hosted,MacOS,X64' --work _work --runasservice
```
In one place it is `x86`, in other `amd64`, here it is `X64`. Can we call them consistently or they have to be called this way?

DaMandal0rian replied:
These are fixed consistently

nazar-pc commented on github-runners/terraform/base/scripts/cleanup.sh:
```diff
@@ -0,0 +1,5 @@
+
+set -euxo pipefail
+
+echo "running clean up script to destroy current deployment environment."
```
It doesn't do anything though :thinking:

nazar-pc commented on github-runners/terraform/base/main.tf:
```diff
@@ -0,0 +1,440 @@
+
+}
+
+resource "aws_instance" "linux_arm_runner" {
```
Similar comments to above apply to all platforms

nazar-pc commented on github-runners/terraform/base/scripts/secure.sh:
```diff
@@ -0,0 +1,249 @@
+#! /bin/bash
```
This is a strange file, looks like it was copy-pasted from somewhere and not really applicable for the most part. Why is it necessary?

nazar-pc reviewed (commented):
Many questions since there wasn't really much explanation why things are done the way are done. Lots of commented-out code that is not clear why it is necessary. Overall please read https://www.notion.so/subspacelabs/Pull-Request-Guide-cd8d0f0f5fe643d5bfde6145450d5ff6

DaMandal0rian committed:
Fixes to deployment

DaMandal0rian committed:
remove k8s runners

DaMandal0rian commented on github-runners/terraform/base/network.tf:
```diff
@@ -0,0 +1,212 @@
+  }
+}
+
+# resource "aws_network_interface" "server_nic" {
```
This was just commented, I left the code here in case we just want to run a single instance and bind the AWS instance NIC to an Elastic IP.

DaMandal0rian commented on github-runners/terraform/base/scripts/secure.sh:
```diff
@@ -0,0 +1,239 @@
+#! /bin/bash
+
```
This script was already modified and tested on Linux OS. It will be further modified with other checks I will include. This will be ported over to a deployment tool eventually.

DaMandal0rian committed:
fix naming conventions, use variables for versions and checksums + minor fixes

- comment unused resources (minor)
- changes to windows runner for rdp

DaMandal0rian committed:
Fix ingress windows

- enable winrm and rdp protocol

DaMandal0rian committed:
Windows instance password and rdp

- remove comments (minor)
- changes to windows provisioners

DaMandal0rian committed:
get runner registration token via api with script

- use actions ci to trigger runner builds
- minor variable adjustments

DaMandal0rian committed:
Minor fixes

- remove packages
- fix arch labels and references
- remove echo from testing
- use encrypted tfvars file as secret

vedhavyas commented on github-runners/terraform/base/main.tf:
```diff
@@ -0,0 +1,380 @@
+  # Github runner installation
+  provisioner "remote-exec" {
+    inline = [
+      "sleep 120",
```
Can you add some explanation on why we need to sleep here ?

DaMandal0rian replied:
It gives time for cloud-init to complete before running remote provisioner

vedhavyas commented on github-runners/terraform/base/main.tf:
```diff
@@ -0,0 +1,380 @@
+}
+
+
+resource "aws_instance" "windows_x86_64_runner" {
```
any plan to caputure the windows metrics through netdata. I see they mentioned an option here - https://blog.netdata.cloud/how-to-monitor-windows-systems-with-netdata/

DaMandal0rian replied:
It's not supported yet in dashboard

vedhavyas commented on github-runners/terraform/base/network.tf:
```diff
@@ -0,0 +1,185 @@
+  }
+
+  ingress {
+    description = "RDP for VPC"
```
why do we need a RDP here and WinRm above?

DaMandal0rian replied:
RDP is for remote access using SSM agent or a RDP client. The WinRM is used for deployments similar to SSH.

vedhavyas commented on github-runners/terraform/base/scripts/cleanup.sh:
```diff
@@ -0,0 +1,10 @@
+./config.sh remove --token ${GH_TOKEN}
+sudo rm -rf /home/ubuntu/actions-runner
+sudo rm -rf /home/ubuntu/_work
+pkill -9 -f actions
```
Sholdn't action be stopped first before pruning the data ?

DaMandal0rian replied:
`./config.sh remove` will handle that.
